### ISSUE TITLE
Partial Serialzer does not work for unassigned transient attributes

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-core#1231]
We implemented a move refactoring using the `IResourceRelocationStrategy` extension point. if we move a single file this works perfectly fine. 

but if we move the declaration and the referencing file at once, then `org.eclipse.xtext.ide.serializer.impl.RelatedResourcesProvider.getRelatedResources(Collection<IResourceSnapshot>)` will remove the related resource from `result ` as both the declaration and the reference are contained in `removed` as both files were moved.

@meysholdt do you remember why we do the sorting out of added/removed.

so if we dont consider moved resources as related who / which code should take care about updating the related resources?

or should `org.eclipse.xtext.ide.serializer.impl.RelatedResourcesProvider.getRelatedResources(Collection<IResourceSnapshot>)` add the new resources to the related resources list too and not the old ones only?

**Update:** as analysis has show the root cause is: Partial Serialzer does not work for unassigned transient attributes

**Comment by cdietrich** [comment]
here is a test showing a similar issue. the workaround that works in our codebase (dont do the remove) does not work there

```
@Test
	def void testMoveToNewPackage2() {
		val fs = new InMemoryURIHandler()
		fs += "inmemory:/file1.fileawaretestlanguage" -> '''
			package pkg1
			
			element Foo {
			}
		'''
		fs += "inmemory:/file2.fileawaretestlanguage" -> '''
			package pkg2
			
			element Bar {
				ref pkg1.Foo
			}
		'''

		val rs = fs.createResourceSet
		val model = rs.contents("inmemory:/file1.fileawaretestlanguage", PackageDeclaration)
		val model2 = rs.contents("inmemory:/file2.fileawaretestlanguage", PackageDeclaration)
		println(model.eResource.errors)
		println(model2.eResource.errors)

		val serializer = serializerProvider.get()
		serializer.addModification(model.eResource)[model.eResource.URI = org.eclipse.emf.common.util.URI.createURI("inmemory:/file1new.fileawaretestlanguage")]
		serializer.addModification(model, [model.name = "newpackage"])
		serializer.addModification(model2.eResource)[model2.eResource.URI = org.eclipse.emf.common.util.URI.createURI("inmemory:/file2new.fileawaretestlanguage")]
		serializer.addModification(model2, [model2.name = "newpackage"])
		Assert.assertEquals(2, model.eResource.resourceSet.resources.size)
		serializer.endRecordChangesToTextDocuments === '''
			--- renamed inmemory:/file1.fileawaretestlanguage to inmemory:/file1new.fileawaretestlanguage (syntax: <offset|text>) 
			package <8:4|newpackage>
			
			element Foo {
			}
			--------------------------------------------------------------------------------
			8 4 "pkg1" -> "newpackage"
			--- renamed inmemory:/file2.fileawaretestlanguage to inmemory:/file2new.fileawaretestlanguage (syntax: <offset|text>) 
			package <8:4|newpackage>
			
			element Bar {
				ref Foo
			}
			--------------------------------------------------------------------------------
			8 4 "pkg2" -> "newpackage"
		'''
	}
```

**Comment by cdietrich** [comment]
it also looks like the scoping in that test behaves differently than it does in our world.
so i dont get any scoping issues if i skipt the remove

**Comment by cdietrich** [comment]
so it looks like all these snapshots cannot cope with the uri change in {{org.eclipse.xtext.ide.refactoring.ResourceRelocationContext.loadAndWatchResource(ResourceRelocationChange)}} as the snapshots all use old uris.

**Comment by cdietrich** [comment]
the problem in the test is that `org.eclipse.xtext.resource.containers.FlatResourceSetBasedAllContainersState` keeps the old data. (installed via ImportTestHelper) it does not seem to be suitable for that usecase.

when removing that the package reference gets updated, but there is still the problem that changes for the old resource are reported (that might have been the reason for removing them in the first place)

```
--- renamed inmemory:/file1.fileawaretestlanguage to inmemory:/file1new.fileawaretestlanguage (syntax: <offset|text>) 
package <8:4|newpackage>

element Foo {
}
--------------------------------------------------------------------------------
8 4 "pkg1" -> "newpackage"
--- renamed inmemory:/file2.fileawaretestlanguage to inmemory:/file2new.fileawaretestlanguage (syntax: <offset|text>) 
package <8:4|newpackage>

element Bar {
	ref <33:8|newpackage.Foo>
}
--------------------------------------------------------------------------------
 8 4 "pkg2" -> "newpackage"
33 8 "pkg1.Foo" -> "newpackage.Foo"
-------- inmemory:/file2.fileawaretestlanguage (syntax: <offset|text>) ---------
package pkg2

element Bar {
	ref <33:8|newpackage.Foo>
}
--------------------------------------------------------------------------------
33 8 "pkg1.Foo" -> "newpackage.Foo"
```

**Comment by cdietrich** [comment]
the test case works with removing the resource though. need to find out why

**Comment by cdietrich** [comment]
```
	@Override
	public void updateReference(ITextRegionDiffBuilder rewriter, IUpdatableReference upd) {
		IUpdatableReference updatable = upd;
		if (rewriter.isModified(updatable.getReferenceRegion())) {
			return;
		}
```

opts out in our case

**Comment by cdietrich** [comment]
here is the trace that creates the rewrite

```
StringBasedTextRegionAccessDiffBuilder.replaceSequence(IHiddenRegion, IHiddenRegion, ISerializationContext, EObject) line: 332	
PartialSerializer$SerializeRecursiveStrategy.serialize(ITextRegionDiffBuilder) line: 160	
PartialSerializer.serializeChanges(ChangeTreeProvider$ResourceRecording, ITextRegionDiffBuilder) line: 260	
RecordingXtextResourceUpdater.applyChange(Deltas, IAcceptor<IEmfResourceChange>) line: 80	
ChangeSerializer.endRecordChanges(IAcceptor<IEmfResourceChange>, IProgressMonitor) line: 191	
ChangeSerializer.applyModifications(IAcceptor<IEmfResourceChange>) line: 113	

```

**Comment by cdietrich** [comment]
`org.eclipse.xtext.ide.serializer.impl.PartialSerializer.trySerializeEObject(EObjectChange, ITextRegionDiffBuilder, SerializationContextMap<IConstraint>)` does seem to not proper handle change to transient features. will double check if we do transient corrent.

`org.eclipse.xtext.ide.serializer.impl.PartialSerializer.trySerializeSingleValue(EObject, FeatureChange, IEObjectRegion, IConstraint)` dislikes transient features without any assignments. it will return null.

**Comment by cdietrich** [comment]
=>

```
public class XXXPartialSerializer extends PartialSerializer {

    @Inject
    private ITransientValueService transientValues;

    @Override
    protected List<SerializationStrategy> trySerializeSingleValue(EObject owner, FeatureChange change,
            IEObjectRegion ownerRegion, IConstraint constraint) {
        List<SerializationStrategy> result = Lists.newArrayList();
        EStructuralFeature feature = change.getFeature();
        IFeatureInfo featureInfo = constraint.getFeatures()[owner.eClass().getFeatureID(feature)];
        IAstRegion featureRegion = findRegion(ownerRegion, change);
        ValueTransient valueTransient = transientValues.isValueTransient(owner, feature);
        if (valueTransient == ValueTransient.YES) {
            return Lists.newArrayList();
        }

        return super.trySerializeSingleValue(owner, change, ownerRegion, constraint);
    }

}
```

@meysholdt @szarnekow any idea why the code checks for assignments first and transiency second?

**Comment by cdietrich** [comment]
https://github.com/eclipse/xtext-core/commit/d3eff0f9753e7ce4d23052ad94edb62c1d6420b6 
and 
https://github.com/eclipse/xtext-core/commit/97191fa5926f25405296021e777472ac69a7cf57

show two variant that should work. @meysholdt what do you think.

unfortuately i have no idea how to test this with a derived metamodel

**Comment by cdietrich** [comment]
fixed in 2.19
